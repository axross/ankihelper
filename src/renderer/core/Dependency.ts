import AnkiService from '../service/AnkiService';
import Backend from '../service/Backend';

type Dependency = {
  ankiService: AnkiService;
  backend: Backend;
};

export default Dependency;
