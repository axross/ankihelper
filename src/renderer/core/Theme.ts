type Theme = {
  text: {
    default: Text;
    headline: Text;
    subtitle: Text;
  };
  spacing: {
    regular: number;
    small: number;
    large: number;
  };
};

type Text = {
  family: string;
  size: number;
  weight: 400 | 500 | 700;
};

export default Theme;
