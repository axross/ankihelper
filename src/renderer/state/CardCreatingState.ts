import { Container } from 'unstated';
import File from '../../common/model/File';

export type State = {
  sessionId: number;
  keyword: string;
  keywordPronunciation: File | null;
  keywordPronunciationLoading: number;
  ipa: string;
  ipaLoading: number;
  partOfSpeech: string;
  definition: string;
  example: string;
  examplePronunciation: File | null;
  examplePronunciationLoading: number;
  exampleWithBlank: string;
  imageSearchQuery: string;
  picture: File | null;
  pictureLoading: number;
  keepExample: boolean;
  cardCreating: number;
};

class CardCreatingState extends Container<State> {
  private readonly defaultState: State;
  public readonly state: State;

  public setKeyword(keyword: string) {
    this.setState({ keyword });
  }

  public setKeywordPronunciation(file: File) {
    this.setState({ keywordPronunciation: file });
  }

  public setIpa(ipa: string) {
    this.setState({ ipa });
  }

  public clearIpa() {
    this.setIpa('');
  }

  public beginLoadingIpa() {
    this.setState(({ ipaLoading }) => ({ ipaLoading: ipaLoading + 1 }));
  }

  public endLoadingIpa() {
    this.setState(({ ipaLoading }) => ({ ipaLoading: ipaLoading - 1 }));
  }

  public setPartOfSpeech(partOfSpeech: string) {
    this.setState({ partOfSpeech });
  }

  public setDefinition(definition: string) {
    this.setState({ definition });
  }

  public setExample(example: string) {
    this.setState({ example });
  }

  public setExamplePronunciation(file: File) {
    this.setState({ examplePronunciation: file });
  }

  public setExampleWithBlank(exampleWithBlank: string) {
    this.setState({ exampleWithBlank });
  }

  public setImageSearchQuery(imageSearchQuery: string) {
    this.setState({ imageSearchQuery });
  }

  public beginLoadingPicture() {
    this.setState(({ pictureLoading }) => ({ pictureLoading: pictureLoading + 1 }));
  }

  public endLoadingPicture() {
    this.setState(({ pictureLoading }) => ({ pictureLoading: pictureLoading - 1 }));
  }

  public setPicture(picture: File) {
    this.setState({ picture });
  }

  public setKeepExample(keepExample: boolean) {
    this.setState({ keepExample });
  }

  public beginCreatingCard() {
    this.setState(({ cardCreating }) => ({ cardCreating: cardCreating + 1 }));
  }

  public endCreatingCard() {
    this.setState(({ cardCreating }) => ({ cardCreating: cardCreating - 1 }));
  }

  public clearAll() {
    const sessionId =
      this.state.sessionId < Number.MAX_SAFE_INTEGER ? this.state.sessionId + 1 : this.defaultState.sessionId;

    this.setState({
      ...this.defaultState,
      sessionId,
      keepExample: this.state.keepExample,
      ...(this.state.keepExample
        ? {
            example: this.state.example,
            examplePronunciation: this.state.examplePronunciation,
            exampleWithBlank: this.state.exampleWithBlank,
            imageSearchQuery: this.state.imageSearchQuery,
            picture: this.state.picture,
          }
        : {}),
    });
  }

  public isReadyToCreate(): boolean {
    if (this.state.keyword === '') return false;
    if (this.state.keywordPronunciation === null) return false;
    if (this.state.ipa === '') return false;
    if (this.state.partOfSpeech === '') return false;
    if (this.state.definition === '') return false;
    if (this.state.example === '') return false;
    if (this.state.examplePronunciation === null) return false;
    if (this.state.exampleWithBlank === '') return false;
    if (this.state.picture === null) return false;
    if (this.state.pictureLoading >= 1) return false;
    if (this.state.cardCreating >= 1) return false;

    return true;
  }

  public constructor() {
    super();

    this.defaultState = {
      sessionId: 1,
      keyword: '',
      keywordPronunciation: null,
      keywordPronunciationLoading: 0,
      ipa: '',
      ipaLoading: 0,
      partOfSpeech: '',
      definition: '',
      example: '',
      examplePronunciation: null,
      examplePronunciationLoading: 0,
      exampleWithBlank: '',
      imageSearchQuery: '',
      picture: null,
      pictureLoading: 0,
      keepExample: false,
      cardCreating: 0,
    };
    this.state = { ...this.defaultState };
  }
}

export default CardCreatingState;
