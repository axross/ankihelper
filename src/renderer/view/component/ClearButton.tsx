import { Button } from '@blueprintjs/core';
import { createElement } from 'react';
import { Subscribe } from 'unstated';
import CardCreatingState from '../../state/CardCreatingState';

type Props = {
  className?: string;
};

const ClearButton = ({ className }: Props) => (
  <Subscribe to={[CardCreatingState]}>
    {(cardCreatingState: CardCreatingState) => (
      <Button onClick={() => cardCreatingState.clearAll()} className={className}>
        Clear
      </Button>
    )}
  </Subscribe>
);

export default ClearButton;
