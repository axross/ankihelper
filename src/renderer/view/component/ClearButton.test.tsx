import { Button } from '@blueprintjs/core';
import 'jest-styled-components';
import { createElement } from 'react';
import * as TestRenderer from 'react-test-renderer';
import { Provider } from 'unstated';
import CardCreatingState from '../../state/CardCreatingState';
import ClearButton from './ClearButton';

describe('<ClearButton>', () => {
  test('matches with the last snapshot', () => {
    const cardCreatingState = new CardCreatingState();

    const testRenderer = TestRenderer.create(
      <Provider inject={[cardCreatingState]}>
        <ClearButton />
      </Provider>
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  test("calls cardCreatingState.clearAll() when Button's props.onClick() is called", () => {
    const cardCreatingState = new CardCreatingState();
    const spy = jest.spyOn(cardCreatingState, 'clearAll');

    const testRenderer = TestRenderer.create(
      <Provider inject={[cardCreatingState]}>
        <ClearButton />
      </Provider>
    );

    testRenderer.root.findByType(Button).props['onClick'](Symbol('event'));

    expect(spy).toHaveBeenCalled();
  });
});
