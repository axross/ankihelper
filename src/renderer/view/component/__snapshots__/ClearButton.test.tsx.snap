// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClearButton> matches with the last snapshot 1`] = `
<button
  className="bp3-button"
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  type="button"
>
  <span
    className="bp3-button-text"
  >
    Clear
  </span>
</button>
`;
